### DISCUSSION TITLE
Sync wave app among users logged in via OIDC

### DISCUSSION THREAD

**Discussion by shamilh2o** [h2oai/wave#1393] [Q&A] [answered]
I need to verify the capability of Wave SDK to sync the wave application among users connected via OIDC. 

I am running a simple wave application with mode = "multicast" and keeping variables in user level. When the wave server runs without OIDC parameters the app, which is running in multiple browsers and browser tabs are syncing as showed in below attachment. 

https://user-images.githubusercontent.com/101506482/165911839-658f8aac-595a-4fd3-809c-5671891b4bd9.mov

When I run the server with OIDC parameters and logged in to the application with multiple users in same time it's not syncing as above. Changes are reflected only for relevant user. 

Are there any suggestions to overcome this and sync the application among all users?

Thanks for the help!
Shamil

**Comment by lo5** [comment]
Thanks for reporting the issue. I've not hit this before. Can you please DM me the OIDC params you're using so that I can try to repro and debug?

**Reply by shamilh2o** [reply]
Yes Sure. I slacked them over to you. Thank you.

**Comment by shamilh2o** [comment]
Any update on this @lo5 ?
cc: @VijithaEkanayake @ozahavi

**Comment by lo5** [comment, marked as answer]
@shamilh2o - I misread your issue as *`multicast` is broken when OIDC is enabled* - sorry :)

`mode=multicast` means "sync the UI across all open tabs of the **same** user". See https://wave.h2o.ai/docs/realtime

So it looks like your app is behaving correctly.

If you want to sync the UI across all open tabs of **all** users, use `mode=broadcast`.

**Reply by shamilh2o** [reply]
Yes @lo5 That's correct. I tried it by using mode='broadcast' too. Actually it synced both users' applications. In our application there is an area to display user name as below. When the mode is broadcast, it also displays same name in both user's applications.

![image](https://user-images.githubusercontent.com/101506482/169741872-32653d00-4229-44d5-903b-6df8799dc47e.png)

That's why I tried to keep mode as 'multicast'.

Note: As I described previously what we are looking is, when user1 updates a table and click a button those table changes should be reflected in the other user's tables without refreshing their application. This happens when we manually refreshes other user's pages. Simply this refresh should be done automatically. 

Thanks in advance.

**Reply by lo5** [reply]
The modes are all-or-nothing. There is no support for using different modes for different parts of the application. I would suggest using unicast mode, and storing the q object for each user in a global dict. When the table is changed, iterate through the dict, change the table, and call q.save() for each q.

**Reply by shamilh2o** [reply]
This solution looks fine @lo5. I tested it with sample application and it worked fine. Will check with our real application. 
Thank you very much for the suggestion.
